addons/mrp/models/mrp_production.py

```diff
@@ -610,54 +608,90 @@ def button_plan(self):
+    def _plan_workorders(self):
+        """ Plan all the production's workorders depending on the workcenters
+        work schedule"""
+        self.ensure_one()
 
         # Schedule all work orders (new ones and those already created)
-        for order in self:
-            start_date = order._get_start_date()
-            from_date_set = False
-            for workorder in order.workorder_ids:
-                workcenter = workorder.workcenter_id
-                wos = WorkOrder.search([('workcenter_id', '=', workcenter.id), ('date_planned_finished', '<>', False),
-                                        ('state', 'in', ('ready', 'pending', 'progress')),
-                                        ('date_planned_finished', '>=', start_date.strftime(DEFAULT_SERVER_DATETIME_FORMAT))], order='date_planned_start')
+        start_date = self._get_start_date()
+        for workorder in self.workorder_ids:
+            workcenters = workorder.workcenter_id | workorder.workcenter_id.alternative_workcenter_ids
+
+            # extract already planned workorders
+            wos = self.env['mrp.workorder'].read_group([
+                ('workcenter_id', 'in', workcenters.ids),
+                ('state', 'in', ('ready', 'pending', 'progress')),
+                ('date_planned_finished', '>=', start_date)],
+                ['workcenter_id', 'date_planned_start:array_agg', 'date_planned_finished:array_agg'],
+                ['workcenter_id'],
+                orderby='date_planned_finished'
+            )
+            wos = {wo['workcenter_id'][0]: zip(wo['date_planned_start'], wo['date_planned_finished']) for wo in wos}
+            best_finished_date = datetime.max
+            best_values = {}
+            for workcenter in workcenters:
+                # get duration expected for this workorder
+                operation = workorder.operation_id
+                cycle_number = float_round(workorder.qty_producing / workcenter.capacity, precision_digits=0, rounding_method='UP')
+                duration_expected = workcenter.time_start + workcenter.time_stop + cycle_number * operation.time_cycle * 100.0 / workcenter.time_efficiency
+
                 from_date = start_date
-                to_date = workcenter.resource_calendar_id.attendance_ids and workcenter.resource_calendar_id.plan_hours(workorder.duration_expected / 60.0, from_date, compute_leaves=True, resource=workcenter.resource_id)
-                if to_date:
-                    if not from_date_set:
-                        # planning 0 hours gives the start of the next attendance
-                        from_date = workcenter.resource_calendar_id.plan_hours(0, from_date, compute_leaves=True, resource=workcenter.resource_id)
-                        from_date_set = True
+                # Check if the workcenter is subject to a schedule
+                has_schedule = workcenter.resource_calendar_id.attendance_ids
```

Comment:
You could create a variable for workcenter_calendar instead of this one, it would be more usefull